src/mlpack/core/data/one_hot_encoding_impl.hpp

```diff
@@ -56,14 +56,141 @@ void OneHotEncoding(const RowType& labelsIn,
   labelMap.clear();
 }
+
+/**
+ * Overloaded function for the above function, which takes a matrix as input
+ * and also a vector of indices to encode and outputs a matrix.
+ * Indices represent the IDs of the dimensions to be one-hot encoded.
+ *
+ * @param input Input dataset to be encoded.
+ * @param indices Index of rows to be encoded.
+ * @param output Encoded matrix.
+ */
+template<typename eT>
+void OneHotEncoding(const arma::Mat<eT>& input,
+                    const arma::Col<size_t>& indices,
+                    arma::Mat<eT>& output)
+{
+  std::unordered_map<size_t, size_t> mapIndices;
+  size_t newRows = input.n_rows - indices.n_elem;
+  for (size_t i = 0; i < indices.n_elem; i++)
+  {
+    mapIndices[indices.at(i)] = numofMapping(input.row(indices.at(i)));
+    newRows += mapIndices[indices.at(i)];
+  }
+
+  // find the size of output mat.
+  output.zeros(newRows, input.n_cols);
+  size_t row = 0;
+
+  for (size_t i = 0; i < input.n_rows; i++)
+  {
+    if (mapIndices.count(i) == 0)
+    {
+      // Copy exactly as required.
+      for (size_t j = 0; j < input.n_cols; j++)
+      {
+        output(row, j) = input(i, j);
+      }
+      row++;
+    }
+    else
+    {
+      arma::Row<size_t> labels;
+      labels.set_size(input.n_cols);
+
+      // Loop over the input labels, and develop the mapping.
+      // Map for labelsIn to labels.
+      std::unordered_map<eT, size_t> labelMap;
+      size_t curLabel = 0;
+      for (size_t k = 0; k < input.n_cols; ++k)
+      {
+        // If labelsIn[i] is already in the map, use the existing label.
+        if (labelMap.count(input(i, k)) != 0)
+        {
+          labels[k] = labelMap[input(i, k)] - 1;
+        }
+        else
+        {
+          // If labelsIn[i] not there then add it to the map.
+          labelMap[input(i, k)] = curLabel + 1;
+          labels[k] = curLabel;
+          ++curLabel;
+        }
+      }
+      // Fill ones in at the required places.
+      for (size_t l = 0; l < input.n_cols; ++l)
+      {
+        output(row + labels[l], l) = 1;
+      }
+      labelMap.clear();
+      row +=  mapIndices[i];
+    }
+  }
+}
+
+/**
+ * Overloaded function for the above function, which takes a matrix as input
+ * and also a DatasetInfo object and outputs a matrix.
+ * This function encodes all the dimensions marked `Datatype::categorical`
+ * in the data::DatasetInfo.
+ *
+ * @param input Input dataset to be encoded.
+ * @param output Encoded matrix.
+ * @param datasetInfo DatasetInfod object that has information about data.
+ */
+template<typename eT>
+void OneHotEncoding(const arma::Mat<eT>& input,
+                    arma::Mat<eT>& output,
+                    const data::DatasetInfo& datasetInfo)
+{
+  std::vector<size_t>indices;
+  for (size_t i = 0; i < datasetInfo.Dimensionality(); ++i)
+  {
+    if (datasetInfo.Type(i) == data::Datatype::categorical)
+    {
+      indices.push_back(i);
+    }
+  }
+  OneHotEncoding(input, (arma::Col<size_t>)(indices), output);
```

Comment:
```suggestion
  OneHotEncoding(input, arma::Col<size_t>(indices), output);
```
I think you don't mean to use a cast here but we can actually call the `arma::Col<size_t>` constructor directly.